ambry-replication/src/main/java/com.github.ambry.replication/ReplicaTokenPersistor.java

```diff
@@ -13,25 +13,202 @@
+   */
+  public final void write(boolean shuttingDown) throws IOException, ReplicationException {
+    for (String mountPath : partitionGroupedByMountPath.keySet()) {
+      write(mountPath, shuttingDown);
+    }
+  }
+
   /**
    * Method to persist the token of partition(s) under the same mountPath.
    * @param mountPath The mouth path of the partition(s).
    * @param shuttingDown indicates whether this is being called as part of shut down.
    */
-  void write(String mountPath, boolean shuttingDown) throws IOException, ReplicationException;
+  public final void write(String mountPath, boolean shuttingDown) throws IOException, ReplicationException {
+    long writeStartTimeMs = SystemTime.getInstance().milliseconds();
+
+    // Get all partitions for the mount path and persist their tokens
+    List<ReplicaTokenInfo> tokensPersisted = new ArrayList<>();
+    for (PartitionInfo info : partitionGroupedByMountPath.get(mountPath)) {
+      for (RemoteReplicaInfo remoteReplica : info.getRemoteReplicaInfos()) {
+        FindToken tokenToPersist = remoteReplica.getTokenToPersist();
+        if (tokenToPersist != null) {
+          tokensPersisted.add(new ReplicaTokenInfo(remoteReplica));
+        }
+      }
+    }
+
+    try {
+      persistTokens(mountPath, tokensPersisted);
+    } finally {
+      // TODO: could move this out
+      replicationMetrics.remoteReplicaTokensPersistTime.update(
+          SystemTime.getInstance().milliseconds() - writeStartTimeMs);
+    }
+
+    for (ReplicaTokenInfo replicaTokenInfo : tokensPersisted) {
+      replicaTokenInfo.getReplicaInfo().onTokenPersisted();
+      if (shuttingDown) {
+        logger.info("Persisted token {}", replicaTokenInfo.getReplicaToken());
+      }
+    }
+  }
 
   /**
-   * Iterates through each mount path and persists all the replica tokens for the partitions on the mount
-   * path to a file. The file is saved on the corresponding mount path.
-   * @param shuttingDown indicates whether this is being called as part of shut down.
+   * Persist the list of tokens in the specified mount path.
+   * @param mountPath
+   * @param tokenInfoList
+   * @throws IOException
    */
-  void write(boolean shuttingDown) throws IOException, ReplicationException;
-}
+  protected abstract void persistTokens(String mountPath, List<ReplicaTokenInfo> tokenInfoList)
+      throws IOException, ReplicationException;
+
+  /**
+   * Retrieve the tokens under the same mountPath.
+   * @param mountPath The mouth path of the partition(s).
+   * @return A list of {@link ReplicaTokenInfo}.
+   */
+  protected abstract List<ReplicaTokenInfo> retrieveTokens(String mountPath) throws IOException, ReplicationException;
+
+  @Override
+  public void run() {
+    try {
+      write(false);
+    } catch (Exception e) {
+      logger.error("Error while persisting the replica tokens {}", e);
+    }
+  }
 
+  /**
+   * Class to serialize and deserialize replica tokens
+   */
+  public static class ReplicaTokenSerde {
```

Comment:
Is it better to move this class to `ReplicaTokeninfo` class?

Reply:
The problem is the serde class requires Clustermap and FindTokenFactory for deser.  It seems weird to construct a ReplicaTokenInfo with those things.

Reply:
Got you. It looks weird.